Chat handling logic improvement + fixes...

Bertram25 opened the pull request:
As requested by @hwoarangmy and to fix a bug seen by @akien-mga .

I also wanted to say that I've been quite ill those few days and I must say I've avoided any kind of screens so my apologize for being invisible the last few days.

But fortunately, I can see it doesn't prevent anyone from bringing on the good stuff. :)

Commit:
Turned the chat and event windows back into file-based layouts + misc requested code fixes.

Commit:
Improved the text queues logic.

- Only updated the widgets texts when necessary (not every frames).
- Fixed scrolling the queues texts with the vertical scrollbar
  as seen by @akien-mga

oyvindln commented on source/modes/GameMode.cpp:
```diff
@@ -1070,63 +1035,93 @@ bool GameMode::keyPressedChat(const OIS::KeyEvent &arg)
+    chatTextBox->appendText(reinterpret_cast<const CEGUI::utf8*>(message->getMessageAsString().c_str()));
+
+    // Ensure the latest text is shown
+    CEGUI::Scrollbar* scrollBar = reinterpret_cast<CEGUI::Scrollbar*>(chatTextBox->getChild("__auto_vscrollbar__"));
```
Are you sure you want to do reinterpret_cast instead of static_cast here? I can see why it's needed for the message text, but don't use reinterpret_cast otherwise unless there is a really good reason for it.

Bertram25 replied:
@oyvindln Unfortunately, the static_casr doesn't compile for me (and for both.).

Bertram25 replied:
@oyvindln But I wanted to say I understand why you'd rather avoid heavier things than static_cast when not necessary. :)

oyvindln replied:
Check again. It should be possible to static_cast a parent class to a child class. Doing that for the scrollbar works fine here.

Bertram25 replied:
Ok, I'll do it now.

Bertram25 replied:
@oyvindln Are you never fed up being right? ;P Static cast are working. I don't know what went through my brains and made me believe it wasn't but well. I'll push a fix.

Commit:
Turned the help window into a file-based layout.

I only left the text within the code as it is the most convenient way
to set it up.

Commit:
Added icons support to event messages.

I had to look in CEGUI code to figure out.

Commit:
Turned useless reinterpret_cast back into static_cast as seen by @oyvindln

Thanks to him for spotting my weird code.

I also use (on purpose) a reinterpret cast for the help text, so that we can
support utf8 text there.

Bertram25 commented:
I'll also work toward making the queues  **not** intercept regular mouse click. Please wait before merging. :)

Commit:
Added more event short notice types.

Also added shot notices for research and objectives.

What's left is to make the chat and event queue not block clicks anymore.

Commit:
Fixed input when clicking in game on the chat or event area.

Bertram25 commented:
It is fixed and is testable IMHO. :D

hwoarangmy commented:
I will test as soon as you release your commit ^^

hwoarangmy commented:
> so my apologize for being invisible the last few days

BTW, I forgot but welcome back ^^

Commit:
Renamed the ignoreInput function as requested by @hwoarangmy :)

Bertram25 commented:
Is done :)

Bertram25 commented:
> BTW, I forgot but welcome back ^^

Thanks :)

hwoarangmy commented:
Testing

hwoarangmy commented:
I confirm the scroll bar works and the icons are cool ^^
Can be merged IMHO

akien-mga commented:
Working fine, I merge.
